pkg/controller/direct/common/hash.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+)
+
+// HashSpec calculates a hash of a k8s spec.
+// We use this to detect changes to the spec.
+func HashSpec(spec interface{}) (string, error) {
+	j, err := json.Marshal(spec)
+	if err != nil {
+		return "", fmt.Errorf("cannot json marshal spec: %w", err)
+	}
+
+	h := sha256.Sum256(j)
+	return hex.EncodeToString(h[:]), nil
+}
+
+// HashProto calculates a hash of a proto message.
+// We use this to detect changes to the GCP resource.
+func HashProto(pb proto.Message) (string, error) {
+	// We normalize the proto by clearing output-only fields etc
+	// We do this on a copy
+	pb = proto.Clone(pb)
+	NormalizeProto(pb)
+
+	// We use a deterministic json marshaler.
+	// We use UseEnumNumbers to ensure that enums are marshaled to their numeric value,
+	// as some APIs (e.g. Run) can return enums as strings ("GA") in some contexts
+	// and as numbers (4) in others. Using the numeric value gives us a stable hash.
+	j, err := protojson.MarshalOptions{UseEnumNumbers: true}.Marshal(pb)
+	if err != nil {
+		return "", fmt.Errorf("cannot json marshal proto: %w", err)
+	}
+
+	h := sha256.Sum256(j)
+	return hex.EncodeToString(h[:]), nil
+}
+
+// NormalizeProto clears fields that are not significant for comparison.
+// It modifies the passed-in proto.
+func NormalizeProto(pb proto.Message) {
+	// TODO: Should we also clear fields like `uid` and `name`?
+	// The problem is that they are not consistently marked as output-only.
+
+	// We clear fields that are known to be volatile
+	volatileFieldNames := []string{"etag", "update_time", "updated_time"}
+
+	// We also clear fields that are marked as output-only
+	// We do this by walking the fields and checking for `field_behavior: OUTPUT_ONLY`
+	// We build a field mask and then use that to clear the fields.
```

Comment:
even if we clear the output-only field, we still update the kcc `.status.observedState` using the current gcp actual proto.